mod/wikirate_source/set/type/source/source.rb

```diff
@@ -35,22 +35,28 @@ def indirect_contributor_search_args
 end
 
 event :process_source_url, :before=>:process_subcards, :on=>:create, :when=>proc{ |c| not (c.subcards["+File"] or c.subcards["+Text"]) } do
-  
+
   linkparams = subcards["+#{ Card[:wikirate_link].name }"]
   url = linkparams && linkparams[:content] or errors.add(:link, " does not exist.")  
-  if url.length != 0 and errors.empty?
-    # errors.add :link, "is empty" 
-  # else
+  if errors.empty? and url.length != 0 
     if Card::Env.params[:sourcebox] == 'true'
-      if cite_card = Card[url] 
+      wikirate_url = "#{ Card::Env[:protocol] }#{ Card::Env[:host] }"
+      if url.start_with?wikirate_url
+        # try to convert the link to source card, easier for users to add source in +source editor
+        uri = URI.parse(URI.unescape(url))
+        cite_card = Card[uri.path] 
+      else
+        cite_card = Card[url] 
+      end
+      if cite_card 
         if cite_card.type_code != :source
           errors.add :source, " can only be source type or valid URL."     
         else
           self.name = cite_card.name
           abort :success
         end
-      else
-        errors.add :source, " does not exist." if not (url.start_with?"http://" or url.start_with?"https://")
+      elsif not (url.start_with?"http://" or url.start_with?"https://") or url.start_with?wikirate_url
```

Comment:
style comment: lack of space after ? is very unusual.

functional comments:
- parens are not necessary.
- start_with? accepts regexp, eg /https?\:\/\//